website/src/_posts/2022-09-3.0.md

```diff
@@ -0,0 +1,146 @@
+import '@uppy/webcam/dist/style.css'
+import '@uppy/audio/dist/style.css'
+import '@uppy/screen-capture/dist/style.css'
+import '@uppy/image-editor/dist/style.css'
+
+// Initialize Uppy
+const uppy = new Uppy()
+  .use(Dashboard, { inline: true, target: '#app' })
+  .use(RemoteSources, { companionUrl: COMPANION_URL, companionAllowedHosts: COMPANION_ALLOWED_HOSTS })
+  .use(Webcam, { target: Dashboard })
+  .use(Audio, { target: Dashboard })
+  .use(ScreenCapture, { target: Dashboard })
+  .use(ImageEditor, { target: Dashboard })
+  .use(Transloadit, {
+    async getAssemblyOptions (file) {
+      // This is where you configure your auth key, auth secret, and template ID
+      // https://uppy.io/docs/transloadit/#getAssemblyOptions-file
+      //
+      // It is important to set the secret in production:
+      // https://transloadit.com/docs/topics/signature-authentication/
+      const response = await fetch('/some-endpoint')
+      return response.json()
+    },
+  })
+```
+
+## Companion 4.0
+
+### Streaming upload
+
+Streaming upload can now be enabled in Companion with Tus. This gives greatly improved upload speeds when using remote Providers and allows uploading potentially hundreds of gigabytes without needing a large server storage.
+
+With streaming upload disabled (default), the whole file will be downloaded first, and only when the download has completely finished would the upload start.
+
+```js
+// Pseudo-code:
+const providerStream = provider.getFileStream()
+const tempFilePath = 'tmpfile'
+const fileStream = fs.createWriteStream(tempFilePath)
+await pipeline(providerStream, fileStream)
+await tus.uploadFile(tempFilePath)
+```
+
+When streaming upload is enabled, Companion will start downloading the file from the provider (for example Google Drive), while at the same time starting the upload to the destination (for example Tus) at the same time, and send every chunk of data consecutively.
+
+```js
+// Pseudo-code:
+const providerStream = provider.getFileStream()
+await tus.uploadStream(providerStream)
+```
+
+For more information, [see the docs](https://uppy.io/docs/companion/).
+
+### Event emitter
+
+Companion now has a server-side event emitter that can be used to detect when uploads start, finish and fail, without having to depend on the client (so it works even if the user has closed their browser).
+
+Example code:
+
+```js
+const { app, emitter } = companion.app(options)
+
+emitter.on('upload-start', ({ token }) => {
+  console.log('Upload started', token)
+
+  function onUploadEvent ({ action, payload }) {
+    if (action === 'success') {
+      emitter.off(token, onUploadEvent) // avoid listener leak
+      console.log('Upload finished', token, payload.url)
+    } else if (action === 'error') {
+      emitter.off(token, onUploadEvent) // avoid listener leak
+      console.error('Upload failed', payload)
+    }
+  }
+  emitter.on(token, onUploadEvent)
+})
+```
+
+### `Async`/`await`-ify Companion
```

Comment:
I think we should combine this with refactor from `request` to `got`, making this more generic (something like "internal improvements").